streaming/base/dataset/streaming.py

```diff
@@ -0,0 +1,623 @@
+                    sleep(TICK)
+                    continue
+
+            # Download and decompress the shard for this sample, if not already done.
+            sample_id = state.sample_ids[state.download_index]
+            shard_id, _ = self.index.find_sample(sample_id)
+            self._download_or_await_shard(shard_states_lock, shard_states, shard_id)
+            state.download_index += 1
+
+    def _each_sample(self, sample_ids: NDArray[np.int64]) -> Iterator[int]:
+        """Iterate over each sample ID, while downloading ahead in the background.
+
+        Args:
+            sample_ids (NDArray[np.int64]): The sample IDs to download and iterate.
+
+        Returns:
+            Iterator[int]: Each sample ID, having been downloaded.
+        """
+        self._partition_state = _PartitionState(sample_ids)
+        Thread(target=self._download_thread, args=(self._partition_state,), daemon=True).start()
+        yield from self._partition_state
+
+    def __iter__(self) -> Iterator[Dict[str, Any]]:
+        """Iterate over all the samples in our partition.
+
+        Returns:
+            Iterator[Dict[str, Any]]: Each sample.
+        """
+        # Exit the thread that is downloading the shards for last epoch, if it exists.
+        if self._partition_state:
+            self._partition_state.stop()
+
+        # Discover where we left off, if there is a checkpoint, or start at the next epoch.
+        # Also pre-increment the epoch counter.
+        world = World()
+        epoch, sample_in_epoch = self._get_progress(world)
+
+        # Get this worker's partition of samples to process.
+        sample_ids = self._get_partition(world, epoch, sample_in_epoch)
+        if sample_ids is None:  # Hit end of epoch, out of samples.
+            return
+
+        # Iterate over the samples while downloading ahead.
+        for sample_id in self._each_sample(sample_ids):
+            yield self[sample_id]
+
+    def state_dict(self, sample_in_epoch: int) -> Dict[str, Any]:
+        """Get a dict containing training state (called from non-worker process).
+
+        This is called on rank zero.
+
+        Args:
+            sample_in_epoch (int): The number of samples processed so far in the current epoch.
+
+        Returns:
+            Dict[str, Any]: The state.
+        """
+        world = World()
+        epoch = self.next_epoch - 1
+        epoch, offset = self._resume(world, epoch)
+        return {
+            'epoch': epoch,
+            'sample_in_epoch': offset + sample_in_epoch,
+        }
+
+    def load_state_dict(self, obj: Dict[str, Any]) -> None:
+        """Load a dict containing training state (called from non-worker process).
+
+        This is called on each copy of the dataset when resuming.
+
+        We just save the state to shared memory for workers to pick up when __iter__ is next
+        called. We use shm because changes to this copy of the dataset wouldn't be picked up by
```

Comment:
can we override this fn and have another variant that takes `epoch`, `sample_in_epoch` as args?